feat: querier test system, ground work

crepererum opened the pull request:
See #3985 for the motivation.

This introduces a cache system for the querier which can later be
extended to support the remaining features listed in #3985 (e.g.
metrics, LRU/TTL).

All current caches are wired up to go throw the new cache system. Once
we move away from (ab)using `db`, the set of caches will be different
but the system will remain.

crepererum commented on querier/src/cache_system/broadcast.rs:
```diff
@@ -0,0 +1,144 @@
+use observability_deps::tracing::warn;
```
This file is copied from `rskafka` where we had a similar application.

crepererum replied:
this file is gone now

crepererum commented on querier/src/cache_system/driver.rs:
```diff
@@ -0,0 +1,430 @@
+use std::{collections::HashMap, future::Future, hash::Hash, sync::Arc};
```
This is the actual cache system.

crepererum replied:
@tustvold I could use your review here.

crepererum commented on querier/src/cache_system/driver.rs:
```diff
@@ -0,0 +1,430 @@
+/// The state parts must be updated in a consistent manner, i.e. while using the same lock guard.
+#[derive(Debug)]
+struct CacheState<K, V> {
+    cached_entries: HashMap<K, V>,
```
Currently we store everything in a "keep all" hash map. That's obviously not the desired end state. I'll introduce an interface for that bit (that also takes care of pruning entries) later, but I didn't want to blow up this PR any further.

crepererum commented on querier/src/cache/mod.rs:
```diff
@@ -0,0 +1,66 @@
+use backoff::BackoffConfig;
```
Everything in `querier::cache` is actually just moved around code. The caches existed beforehand but in a giant struct and hand-managed hashmaps. They are not split apart. The tests are the same (just moved to the corresponding modules).

crepererum commented:
FYI: @tustvold proposed to just use a shared oneshot channel instead of `BroadcastOnce`. This would also avoid the usage of the weird panic guard. I'll implement that tomorrow. This shouldn't affect the interface or the tests though.

tustvold commented:
Looks good to me, nice work. As you've articulated, whilst I like the idea of reusing BroadcastOnce, I think the fact we don't need peek() functionality and already have external state storage means we can do something simpler

crepererum commented on querier/src/cache_system/driver.rs:
```diff
@@ -0,0 +1,430 @@
+    /// Flag that the loader did not panic.
+    ///
+    /// Dropping this guard after calling this function will NOT broadcast anything.
+    fn did_not_panic(&mut self) {
```
this code is gone now

crepererum commented on querier/src/cache_system/driver.rs:
```diff
@@ -0,0 +1,430 @@
+    /// Create new, empty cache with given loader function.
+    pub fn new<T, F>(loader: T) -> Self
+    where
+        T: Fn(K) -> F + Send + Sync + 'static,
```
done

crepererum commented on querier/src/cache_system/driver.rs:
```diff
@@ -0,0 +1,430 @@
+    fn setup() -> (Arc<Cache<u8, String>>, Arc<Loader>) {
+        let loader = Arc::new(Loader::default());
+        let loader_captured = Arc::clone(&loader);
+        let cache = Arc::new(Cache::new(move |key| {
```
done

crepererum commented:
@tustvold I've removed `BroadcastOnce`, which resulted in way simpler and far less code. I've also introduced a `Loader` trait. So PTAL :slightly_smiling_face:

crepererum commented on querier/src/cache_system/driver.rs:
```diff
@@ -0,0 +1,391 @@
+
+                    // remove "running" state and store result
+                    let mut state = state_captured.lock();
+                    drop(
```
I'll add a comment. In short: to convince `rustc` that we don't need the shared future that is returned by `remove`.
